pkg/rpc/response/result/version.go

```diff
@@ -32,8 +65,98 @@ type (
+		Network                     netmode.Magic   `json:"network"`
+		MillisecondsPerBlock        int             `json:"msperblock"`
+		MaxTraceableBlocks          uint32          `json:"maxtraceableblocks"`
+		MaxValidUntilBlockIncrement uint32          `json:"maxvaliduntilblockincrement"`
+		MaxTransactionsPerBlock     uint16          `json:"maxtransactionsperblock"`
+		MemoryPoolMaxTransactions   int             `json:"memorypoolmaxtransactions"`
+		ValidatorsCount             byte            `json:"validatorscount"`
+		InitialGasDistribution      json.RawMessage `json:"initialgasdistribution"`
+		StateRootInHeader           bool            `json:"staterootinheader,omitempty"`
 	}
 )
+
+// MarshalJSON implements the json marshaller interface.
+func (v *Version) MarshalJSON() ([]byte, error) {
+	aux := versionMarshallerAux{
+		Magic:     v.Magic,
+		TCPPort:   v.TCPPort,
+		WSPort:    v.WSPort,
+		Nonce:     v.Nonce,
+		UserAgent: v.UserAgent,
+		Protocol: protocolMarshallerAux{
+			AddressVersion:              v.Protocol.AddressVersion,
+			Network:                     v.Protocol.Network,
+			MillisecondsPerBlock:        v.Protocol.MillisecondsPerBlock,
+			MaxTraceableBlocks:          v.Protocol.MaxTraceableBlocks,
+			MaxValidUntilBlockIncrement: v.Protocol.MaxValidUntilBlockIncrement,
+			MaxTransactionsPerBlock:     v.Protocol.MaxTransactionsPerBlock,
+			MemoryPoolMaxTransactions:   v.Protocol.MemoryPoolMaxTransactions,
+			ValidatorsCount:             v.Protocol.ValidatorsCount,
+			InitialGasDistribution:      int64(v.Protocol.InitialGasDistribution),
+			StateRootInHeader:           v.Protocol.StateRootInHeader,
+		},
+		StateRootInHeader: v.StateRootInHeader,
+	}
+	return json.Marshal(aux)
+}
+
+// UnmarshalJSON implements the json unmarshaller interface.
+func (v *Version) UnmarshalJSON(data []byte) error {
+	var aux versionUnmarshallerAux
+	err := json.Unmarshal(data, &aux)
+	if err != nil {
+		return err
+	}
+	v.Magic = aux.Magic
+	v.TCPPort = aux.TCPPort
+	v.WSPort = aux.WSPort
+	v.Nonce = aux.Nonce
+	v.UserAgent = aux.UserAgent
+	v.Protocol.AddressVersion = aux.Protocol.AddressVersion
+	v.Protocol.Network = aux.Protocol.Network
+	v.Protocol.MillisecondsPerBlock = aux.Protocol.MillisecondsPerBlock
+	v.Protocol.MaxTraceableBlocks = aux.Protocol.MaxTraceableBlocks
+	v.Protocol.MaxValidUntilBlockIncrement = aux.Protocol.MaxValidUntilBlockIncrement
+	v.Protocol.MaxTransactionsPerBlock = aux.Protocol.MaxTransactionsPerBlock
+	v.Protocol.MemoryPoolMaxTransactions = aux.Protocol.MemoryPoolMaxTransactions
+	v.Protocol.ValidatorsCount = aux.Protocol.ValidatorsCount
+	v.Protocol.StateRootInHeader = aux.Protocol.StateRootInHeader
+	v.StateRootInHeader = aux.StateRootInHeader
+	if len(aux.Protocol.InitialGasDistribution) == 0 {
+		return nil
+	}
+	if strings.HasPrefix(v.UserAgent, config.UserAgentPrefix) && strings.Compare(v.UserAgent, fmt.Sprintf(config.UserAgentFormat, "0.98.3")) < 0 {
```

Comment:
`strings.Compare` is not exactly correct for this.

Reply:
I'd suggest to compare full git revisions instead of "0.98.3", but it's hard to implement. What should we use here?

Reply:
https://github.com/coreos/go-semver
https://github.com/hashicorp/go-version

Not sure which is the best one, but something like this should work.

Reply:
Wow, nice tools, I wasn't aware that something like this exists.

Reply:
Fixed.